server/src/postgres/simple_resource.rs

```diff
@@ -0,0 +1,404 @@
+#[tonic::async_trait]
+pub trait PsqlObjectType<T>
+where
+    Self: PsqlType + ObjectType<T> + Send + SimpleResource<T>,
+    T: GrpcDataObjectType,
+{
+    /// get data from the database using the Object's UUID
+    /// returns [Row] on success
+    async fn read(&self) -> Result<Row, ArrErr> {
+        //TODO: implement shared memcache here to get object data if present
+        let id = self.try_get_uuid()?;
+        Self::get_by_id(&id).await
+    }
+
+    /// update the Object's database record using provided data
+    /// returns [Option(Row)] and [ValidationResult]
+    /// returns [ArrErr] if any error is thrown
+    async fn update<'a>(&self, data: &T) -> Result<(Option<Row>, ValidationResult), ArrErr> {
+        let (psql_data, validation_result) = Self::validate(data)?;
+
+        if !validation_result.success {
+            return Ok((None, validation_result));
+        }
+
+        let definition = Self::get_definition();
+        let id_col = Self::try_get_id_field()?;
+        let id = self.try_get_uuid()?;
+        let mut params: Vec<&PsqlField> = vec![];
+        let mut updates = vec![];
+        let mut index = 1;
+
+        for key in definition.fields.keys() {
+            match psql_data.get(&*key.to_string()) {
+                Some(value) => {
+                    let val: &PsqlField = <&Box<PsqlFieldSend>>::clone(&value).as_ref();
+                    updates.push(format!("{} = ${}", key, index));
+                    params.push(val);
+                    index += 1;
+                }
+                None => {
+                    psql_debug!(
+                        "Skipping update [{}] for [{}] with id [{}]",
+                        key,
+                        definition.psql_table,
+                        id
+                    );
+                }
+            }
+        }
+
+        let update_sql = &format!(
+            "UPDATE {} SET {} WHERE {} = ${}",
+            definition.psql_table,
+            updates.join(", "),
+            id_col,
+            index
+        );
+        psql_debug!("{}", update_sql);
+        params.push(&id);
+        psql_debug!("{:?}", &params);
+
+        psql_info!(
+            "Updating entry in table [{}]. uuid: {}",
+            definition.psql_table,
+            id
+        );
+        let client = get_psql_pool().get().await?;
+        client.execute(update_sql, &params[..]).await?;
+
+        //TODO: flush shared memcache for this resource when memcache is implemented
+
+        Ok((Some(self.read().await?), validation_result))
+    }
+
+    /// calls [set_deleted_at_now](PsqlObjectType::set_deleted_at_now) if the Object has a `deleted_at` field
+    /// calls [delete_row](PsqlObjectType::delete_row) otherwise
+    async fn delete(&self) -> Result<(), ArrErr> {
+        let definition = Self::get_definition();
+        if definition.fields.contains_key("deleted_at") {
+            self.set_deleted_at_now().await
```

Comment:
Overwriting an older deletion time with a newer one could raise some debug and auditing issues, would rather ignore if it's already set and issue a log message

Reply:
That's a good point! Will add a check if it's already set before updating.

Reply:
Added a `is_archived` function to check if the `deleted_at` field is already set for the resource.
https://github.com/Arrow-air/svc-storage/blob/6883178b0d80010b58ae192fab6479d96484803c/server/src/postgres/simple_resource.rs#L332

This is now being used to check if we want to update the field or not:
https://github.com/Arrow-air/svc-storage/blob/6883178b0d80010b58ae192fab6479d96484803c/server/src/postgres/simple_resource.rs#L372